cranelift/codegen/src/isa/riscv64/inst_vector.isle

```diff
@@ -254,3 +306,70 @@
+
+;; Helper for emitting the `vslidedown.vx` instruction.
+;; `vslidedown` moves all elements in the vector down by n elements.
+;; The top most elements are up to the tail policy.
+(decl rv_vslidedown_vx (Reg Reg VState) Reg)
+(rule (rv_vslidedown_vx vs2 vs1 vstate)
+  (vec_alu_rrr (VecAluOpRRR.VslidedownVX) vs2 vs1 vstate))
+
+;; Helper for emitting the `vslidedown.vi` instruction.
+;; Unlike other `vi` instructions the immediate is zero extended.
+(decl rv_vslidedown_vi (Reg UImm5 VState) Reg)
+(rule (rv_vslidedown_vi vs2 imm vstate)
+  (vec_alu_rr_uimm5 (VecAluOpRRImm5.VslidedownVI) vs2 imm vstate))
+
+;; Helper for emitting the `vmv.x.s` instruction.
+;; This instruction copies the first element of the source vector to the destination X register.
+(decl rv_vmv_xs (Reg VState) Reg)
+(rule (rv_vmv_xs vs vstate)
+  (vec_alu_rr (VecAluOpRR.VmvXS) vs vstate))
+
+;; Helper for emitting the `vfmv.f.s` instruction.
+;; This instruction copies the first element of the source vector to the destination F register.
+(decl rv_vfmv_fs (Reg VState) Reg)
+(rule (rv_vfmv_fs vs vstate)
+  (vec_alu_rr (VecAluOpRR.VfmvFS) vs vstate))
+
+;; Helper for emitting the `vmv.v.x` instruction.
+;; This instruction splats the X regsiter into all elements of the destination vector.
+(decl rv_vmv_vx (Reg VState) Reg)
+(rule (rv_vmv_vx vs vstate)
+  (vec_alu_rr (VecAluOpRR.VmvVX) vs vstate))
+
+;; Helper for emitting the `vfmv.v.f` instruction.
+;; This instruction splats the F regsiter into all elements of the destination vector.
+(decl rv_vfmv_vf (Reg VState) Reg)
+(rule (rv_vfmv_vf vs vstate)
+  (vec_alu_rr (VecAluOpRR.VfmvVF) vs vstate))
+
+;; Helper for emitting the `vmv.v.i` instruction.
+;; This instruction splat's the immediate value into all elements of the destination vector.
+(decl rv_vmv_vi (Imm5 VState) Reg)
+(rule (rv_vmv_vi imm vstate)
+  (vec_alu_r_imm5 (VecAluOpRImm5.VmvVI) imm vstate))
+
+;;;; Multi-Instruction Helpers ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(decl gen_extractlane (Type Reg u8) Reg)
+
+;; When extracting lane 0 for floats, we can use `vfmv.f.s` directly.
+(rule 3 (gen_extractlane (ty_vec_fits_in_register ty) src 0)
+  (if (ty_vector_float ty))
+  (rv_vfmv_fs src ty))
+
+;; When extracting lane 0 for integers, we can use `vmv.x.s` directly.
+(rule 2 (gen_extractlane (ty_vec_fits_in_register ty) src 0)
+  (if (ty_vector_not_float ty))
+  (rv_vmv_xs src ty))
+
+;; In the general case, we must first use a `vslidedown` to place the correct lane
+;; in index 0, and then use the appropriate `vmv` instruction.
+;; If the index fits into a 5-bit immediate, we can emit a `vslidedown.vi`.
+(rule 1 (gen_extractlane (ty_vec_fits_in_register ty) src (uimm5_from_u8 idx))
+  (gen_extractlane ty (rv_vslidedown_vi src idx ty) 0))
+
+;; Otherwise lower it into an X register.
+(rule 0 (gen_extractlane (ty_vec_fits_in_register ty) src idx)
+  (gen_extractlane ty (rv_vslidedown_vx src (imm $I64 idx) ty) 0))
```

Comment:
Mostly to avoid having two rules for float and integer in the slide down cases. `vslidedown` is generic across integer and float elements, but `vmv` is not, so we recursively call this rule to decide the correct `vmv` instruction to use.

That being said, we can probably inline the `vslidedown` rules and have just a generic `vmv` that decides the correct instruction based on the type.

Reply:
Oh sorry to clarify I mean that you've got 4 cases of `gen_extractlane` here, but why not have those 4 cases be cases on `lower (extractlane ..)`?

Reply:
If I inline it directly into `(lower (extractlane ..))` then it would become, 6 rules, right? Since I would have to duplicate `vslidedown.vi+vmv`, `vslidedown.vi+vfm`, `vslidedown.vx+vmv` and `vslidedown.vx+vfm`,  which are currently just 2 rules.

Or can I then recursively call the lower instructions directly?

Reply:
Ah I apologize I missed that crucial bit of this being a recursive rule! In that case yeah definitely makes sense as a standalone decl.